File: two_sum/twoSum.js

function violence(nums, target) {
    // 暴力解法
    for (i = 0; i < nums.length - 1; i++) {
        for (j = i + 1; j < nums.length; j++) {
            let complement = target - nums[i]
            if (nums[j] == complement) {
                return [i, j]
            }
        }
    }
    throw Error("Can't find solution!")
}

function twoHashTraversing(nums, target) {
    // 两遍哈希
    let hashMap = {}
    nums.forEach((value, index) => {
        hashMap[value] = index
    });

    for (index in nums) {
        let complement = target - nums[index]
        if (hashMap[complement] != void 0 && hashMap[complement] != index) {
            return [Number(index), hashMap[complement]]
        }
    }
    throw Error("Can't find solution!")
}

function oneHashTraversing(nums, target) {
    // 一遍哈希
    let hashMap = {}
    for (index in nums) {
        let complement = target - nums[index]
        if (hashMap[complement] != void 0) {
            return [Number(hashMap[complement]), Number(index)]
        }
        hashMap[nums[index]] = index
    }
    throw Error("Can't find solution!")
}

let numbers = [2,7,11,15]
let result = 9

console.log("numbers is %s, result is %d", numbers, result)
console.log("violence output index is %s", violence(numbers, result))
console.log("twoHashTraversing output index is %s", twoHashTraversing(numbers, result))
console.log("oneHashTraversing output index is %s", oneHashTraversing(numbers, result))

File: two_sum/twoSum.py
#!/usr/local/bin/python3
#coding=utf-8

from typing import List

class Solution:
    def violence(self, nums: List[int], target: int) -> List[int]:
        """
        :暴力解法 (letcode超时)
        """
        i = 0
        while i < len(nums) - 1:
            j = i + 1
            while j < len(nums):
                complement = target - nums[i]
                if complement == nums[j]:
                    return [i, j]
                j += 1
            i += 1
        raise ValueError("Can't find solution!")

    def violence_2(self, nums: List[int], target: int) -> List[int]:
        """
        :暴力解法2
        """
        for i in range(0, len(nums) - 1):
            for j in range(i + 1, len(nums)):
                complement = target - nums[i]
                if complement == nums[j]:
                    return [i, j]

        raise ValueError("Can't find solution!")


    def two_hash_traversing(self, nums: List[int], target: int) -> List[int]:
        """
        :两遍哈希表
        """
        hashmap = {}
        for index in range(len(nums)):
            hashmap[nums[index]] = index
        
        for index, value in enumerate(nums):
            complement = target - value
            if complement in hashmap and hashmap[complement] != index:
                return [index, hashmap[complement]]

        raise ValueError("Can't find solution!")

    def one_hash_traversing(self, nums: List[int], target: int) -> List[int]:
        """
        :一遍哈希表
        """
        hashmap = {}
        for index, value in enumerate(nums):
            complement = target - value
            if complement in hashmap:
                return [hashmap[complement], index]
            hashmap[value] = index

        raise ValueError("Can't find solution!")

def main():
    numbers = [2,7,11,15]
    result = 9

    solution = Solution()
    print("numbers is %s, result is %d" % (numbers, result))
    print("violence output index is %s" % (solution.violence(numbers, result)))
    print("violence_2 output index is %s" % (solution.violence_2(numbers, result)))
    print("two_hash_traversing output index is %s" % (solution.two_hash_traversing(numbers, result)))
    print("one_hash_traversing output index is %s" % (solution.one_hash_traversing(numbers, result)))

if __name__ == "__main__":
    main()